util/gpool/spmc/spmspool.go

```diff
@@ -0,0 +1,403 @@
+func (p *Pool[T, U, C, CT, TF]) AddProduceBySlice(producer func() ([]T, error), constArg C, contextFn TF, options ...TaskOption) (<-chan U, pooltask.TaskController[T, U, C, CT, TF]) {
+	opt := loadTaskOptions(options...)
+	taskID := p.NewTaskID()
+	var wg sync.WaitGroup
+	result := make(chan U, opt.ResultChanLen)
+	closeCh := make(chan struct{})
+	tc := pooltask.NewTaskController[T, U, C, CT, TF](p, taskID, closeCh, &wg)
+	taskCh := make(chan pooltask.Task[T], opt.TaskChanLen)
+	for i := 0; i < opt.Concurrency; i++ {
+		err := p.run()
+		if err == gpool.ErrPoolClosed {
+			break
+		}
+		taskBox := pooltask.NewTaskBox[T, U, C, CT, TF](constArg, contextFn, &wg, taskCh, result, taskID)
+		p.taskCh <- &taskBox
+	}
+	go func() {
+		defer func() {
+			if r := recover(); r != nil {
+				logutil.BgLogger().Error("producer panic", zap.Any("recover", r), zap.Stack("stack"))
+			}
+			close(closeCh)
+			close(taskCh)
+		}()
+		for {
+			tasks, err := producer()
+			if err != nil {
+				if errors.Is(err, gpool.ErrProducerClosed) {
+					return
+				}
+				log.Error("producer error", zap.Error(err))
+				return
+			}
+			for _, task := range tasks {
+				wg.Add(1)
+				task := pooltask.Task[T]{
+					Task: task,
+				}
+				taskCh <- task
+			}
+		}
+	}()
+	return result, tc
+}
+
+// AddProducer is to add producer.
+// Producer need to return ErrProducerClosed when to exit.
+func (p *Pool[T, U, C, CT, TF]) AddProducer(producer func() (T, error), constArg C, contextFn TF, options ...TaskOption) (<-chan U, pooltask.TaskController[T, U, C, CT, TF]) {
+	opt := loadTaskOptions(options...)
+	taskID := p.NewTaskID()
+	var wg sync.WaitGroup
+	result := make(chan U, opt.ResultChanLen)
+	closeCh := make(chan struct{})
+	tc := pooltask.NewTaskController[T, U, C, CT, TF](p, taskID, closeCh, &wg)
+	taskCh := make(chan pooltask.Task[T], opt.TaskChanLen)
+	for i := 0; i < opt.Concurrency; i++ {
+		err := p.run()
+		if err == gpool.ErrPoolClosed {
+			break
+		}
+		taskBox := pooltask.NewTaskBox[T, U, C, CT, TF](constArg, contextFn, &wg, taskCh, result, taskID)
+		p.taskCh <- &taskBox
+	}
+	go func() {
+		defer func() {
+			if r := recover(); r != nil {
+				logutil.BgLogger().Error("producer panic", zap.Any("recover", r), zap.Stack("stack"))
+			}
+			close(closeCh)
+			close(taskCh)
+		}()
+		for {
+			task, err := producer()
+			if err != nil {
+				if errors.Is(err, gpool.ErrProducerClosed) {
+					return
+				}
+				log.Error("producer error", zap.Error(err))
+				return
+			}
+			wg.Add(1)
+			t := pooltask.Task[T]{
+				Task: task,
+			}
+			taskCh <- t
+		}
+	}()
+	return result, tc
+}
+
+func (p *Pool[T, U, C, CT, TF]) run() error {
+	if p.IsClosed() {
+		return gpool.ErrPoolClosed
+	}
+	var w *goWorker[T, U, C, CT, TF]
+	if w = p.retrieveWorker(); w == nil {
+		return gpool.ErrPoolOverload
+	}
+	return nil
+}
+
+// retrieveWorker returns an available worker to run the tasks.
+func (p *Pool[T, U, C, CT, TF]) retrieveWorker() (w *goWorker[T, U, C, CT, TF]) {
```

Comment:
The function seems to be the same with `github.com/ngaut/pools`? This pkg has been used at multiple places in TiDB. Could we use it directly rather than inverting a new wheel?

1. Get / Put, maintain the initial size and capacity
2. Auto scaling and shrinking the workers (resource) according to expiry.

Does this pool / worker_loop_queue provide more function?

Reply:
In fact, They are two different things. ```ngaut/pools``` is a object pool. ```gpool``` is a goroutine pool. ```ngaut/pools``` is not a good implement. Someone wants to refactor it
https://github.com/pingcap/tidb/pull/32878

Reply:
this is very strange.```ngaut/pools``` is an object pool. but it also can control app concurrency. I think if you want to use an object pool, sync.Pool is enough.

Reply:
> Does this pool / worker_loop_queue provide more function?

Yes, the worker loop queue is one of implemented. in fact, we have an implement by stack.